File: leetcode/two_pointers/python/125_valid_palindrome.py
# A phrase is a palindrome if, after converting all uppercase letters into lowercase letters
# and removing all non-alphanumeric characters, it reads the same forward and backward.
# Alphanumeric characters include letters and numbers.
# Given a string s, return true if it is a palindrome, or false otherwise.
# Example 1:
# Input: s = "A man, a plan, a canal: Panama"
# Output: true
# Explanation: "amanaplanacanalpanama" is a palindrome.

class Solution:
    def check(self, s: str, begin: int, end: int) -> bool:
        # If the begin pointer is not character, recur +1 begin
        # if the end pointer is not a character, recur -1 end
        # else if lower case of begin and end doesn't match return false
        # recur +1 begin -1 end
        if begin >= end:
            return(True)
        if s[begin].isalpha() == False:
            return self.check(s, begin + 1, end)
        elif s[end].isalpha() == False:
            return self.check(s, begin, end-1)
        else:
            b = s[begin].lower()
            e = s[end].lower()
            if b != e:
                return False
            return self.check(s, begin + 1, end -1)


    def isPalindrome(self, s: str) -> bool:
        # Recursively check if begin and end pointer are the same
        # if not return false
        return self.check(s, 0, len(s)-1)

File: leetcode/arrays_and_hashing/golang/two_pointers/125_valid_palindrome.go
/*
Is Palindrome

Given a string s, return true if it is a palindrome, otherwise return false.

A palindrome is a string that reads the same forward and backward. It is also case-insensitive and ignores all non-alphanumeric characters.

Example 1:

Input: s = "Was it a car or a cat I saw?"

Output: true

Explanation: After considering only alphanumerical characters we have "wasitacaroracatisaw", which is a palindrome.

Example 2:

Input: s = "tab a cat"

Output: false

Explanation: "tabacat" is not a palindrome.

Constraints:

    1 <= s.length <= 1000
    s is made up of only printable ASCII characters.

*/

// Time: O(n)
// Space: O(1)
// Two Pointers Solution
package main

import "strings"

func isPalindrome(s string) bool {
	// Convert the string to lowercase
	s = strings.ToLower(s)
	// Initialize the left and right pointers
	left, right := 0, len(s)-1
	// Iterate until the left pointer is less than the right pointer
	for left < right {
		// If the characters at the left and right pointers are not alphanumeric, move the pointers
		if !isAlphanumeric(s[left]) {
			left++
		} else if !isAlphanumeric(s[right]) {
			right--
		} else {
			// If the characters at the left and right pointers are alphanumeric, check if they are equal
			if s[left] != s[right] {
				return false
			}
			// Move the pointers
			left++
			right--
		}
	}
	// If the string is a palindrome, return true
	return true
}

func isAlphanumeric(b byte) bool {
	return (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || (b >= '0' && b <= '9')
}
